core/sile.lua

```diff
@@ -240,10 +240,59 @@ function SILE.resolveFile(filename, pathprefix)
   return nil
 end
 
+SILE.currentCommandStack = {}
+--- Push currently processed command on to the stack to record the execution trace for debugging.
+function SILE.currentCommandStack.push(stack, command)
+  if command == nil then return end
+
+  stack[#stack + 1] = command
+  SILE.currentCommand = command
+end
+--- Pop previously pushed command from the stack.
+--- Command is provided to check that pushes and pops are balanced.
+function SILE.currentCommandStack.pop(stack, command)
+  if command == nil then return end
+
+  local popped = stack[#stack]
+  if popped ~= command then
+    SU.warn("Unbalanced command push/pop. Expected '" .. popped .. "', got '" .. command .. "'.")
+    io.stderr:write(debug.traceback())
+    io.stderr:write("\n")
+
+    os.exit(1)
+  else
+    stack[#stack] = nil
+  end
+end
+--- Returns a stack of commands, which can be used to show information about location in document.
+--- Most relevant object is in the last index.
+--- Second return value is true, if the most relevant object is no longer processed (has been popped)
+function SILE.currentCommandStack.locationStack(stack)
+  local result = {}
+
+  for i = 1, #stack do
+    if type(stack[i]) == "table" and stack[i].line then
+      result[#result + 1] = stack[i]
+    end
+  end
+
+  local lastIsExtra = false
+
+  if type(SILE.currentCommand) == "table" and SILE.currentCommand.line and result[#result] ~= SILE.currentCommand then
+    result[#result + 1] = SILE.currentCommand
+    lastIsExtra = true
+  end
+
+  return result, lastIsExtra
+end
+
 function SILE.call(cmd, options, content)
-  SILE.currentCommand = content
+  SILE.currentCommandStack:push(content)
```

Comment:
Unfortunately `SILE.call()` is not a reliable way to hook into functions being called. They can (and even in SILE's own code, do) get called directly from `SILE.Commands["cmd"]()`.

Reply:
I agree, I haven't realized that at the time of writing it.

Reply:
I think that’s a bug (or at least a historical artefact) if they do, though. We should be able to assume SILE.call, especially if it makes this kind of hooking easier.

Reply:
Working on fixing this, see #605.